File: kotlin/shared/tests/com/xplpc/library/DataListTest.kt
package com.xplpc.library

import com.xplpc.data.CallbackList
import com.xplpc.data.MappingList
import com.xplpc.helper.ByteArrayHelper
import com.xplpc.map.MappingItem
import com.xplpc.message.Message
import com.xplpc.type.DataView
import java.nio.ByteBuffer
import java.util.concurrent.atomic.AtomicInteger
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull

class DataListTest {
    private fun setUp() {
        TestPlatform.initialize()
    }

    @AfterTest
    fun tearDown() {
        CallbackList.clear()
    }

    @Test
    fun callbackListExecutesOnlyOnce() {
        setUp()

        var calls = 0

        CallbackList.add("callback-once") { calls += 1 }

        CallbackList.execute("callback-once", "")
        CallbackList.execute("callback-once", "")

        assertEquals(1, calls)
    }

    @Test
    fun callbackListRemovesPendingEntry() {
        setUp()

        var calls = 0

        CallbackList.add("callback-removed") { calls += 1 }
        CallbackList.remove("callback-removed")
        CallbackList.execute("callback-removed", "")

        assertEquals(0, calls)
        assertEquals(0, CallbackList.count())
    }

    @Test
    fun callbackListIsTakenByASingleThread() {
        setUp()

        val calls = AtomicInteger()

        CallbackList.add("callback-concurrent") { calls.incrementAndGet() }

        val threads = (1..16).map {
            Thread { CallbackList.execute("callback-concurrent", "") }
        }

        threads.forEach { it.start() }
        threads.forEach { it.join() }

        assertEquals(1, calls.get())
    }

    @Test
    fun mappingListSurvivesClear() {
        setUp()

        MappingList.add("sample.temporary", MappingItem { _, r -> r(null) })
        MappingList.clear()

        assertEquals(0, MappingList.count())
        assertFalse(MappingList.has("sample.temporary"))
    }

    @Test
    fun dataViewFromByteArraySurvivesGarbageCollection() {
        setUp()

        val data = byteArrayOf(1, 2, 3, 4, 5, 6, 7, 8)
        val dataView = DataView.createFromByteArray(data)

        // The source array is dropped and the heap is churned, and the view must still address its own storage.
        System.gc()
        System.gc()

        repeat(64) { ByteArray(64 * 1024) }

        assertEquals(8, dataView.size)
        assertEquals(data.toList(), ByteArrayHelper.createFromDataView(dataView).toList())
    }

    @Test
    fun messageValueReturnsNullForAnotherType() {
        val message = Message()
        message.set("number", 10)

        assertEquals(10, message.get<Int>("number"))
        assertNull(message.get<String>("number"))
        assertNull(message.get<Int>("missing"))
    }

    // A view that carries no address describes nothing to read, and reading it must not fault.
    @Test
    fun emptyDataViewReadsAsNothing() {
        assertEquals(0, ByteArrayHelper.createFromDataView(DataView(0, 16)).size)
        assertEquals(0, ByteArrayHelper.createFromDataView(DataView(128, 0)).size)
        assertEquals(0, ByteArrayHelper.createFromDataView(DataView(128, -1)).size)
    }

    // A heap buffer has no address native code can use, so the view has to refuse it rather than point at zero.
    @Test
    fun heapBufferIsRefused() {
        val view = DataView.createFromByteBuffer(ByteBuffer.allocate(16))

        assertEquals(0L, view.ptr)
        assertEquals(0, view.size)
    }
}

File: flutter/plugin/test/data_list_test.dart
import 'package:logging/logging.dart';
import 'package:test/test.dart';
import 'package:xplpc/core/config.dart';
import 'package:xplpc/core/xplpc.dart';
import 'package:xplpc/data/callback_list.dart';
import 'package:xplpc/helper/byte_array_helper.dart';
import 'package:xplpc/data/mapping_list.dart';
import 'package:xplpc/map/mapping_item.dart';
import 'package:xplpc/message/message.dart';
import 'package:xplpc/type/dataview.dart';
import 'package:xplpc/serializer/json_serializer.dart';
import 'package:xplpc/util/unique_id.dart';

void main() {
  group('Testing Data Lists', () {
    setUpAll(() {
      XPLPC.instance.initialize(Config(JsonSerializer()));
    });

    tearDown(() {
      CallbackList.instance.clear();
      MappingList.instance.clear();
    });

    test('Callback List Executes Only Once', () {
      var calls = 0;

      CallbackList.instance.add("callback-once", (_) => calls += 1);

      CallbackList.instance.execute("callback-once", "");
      CallbackList.instance.execute("callback-once", "");

      expect(calls, 1);
    });

    test('Callback List Registers Synchronously', () {
      CallbackList.instance.add("callback-sync", (_) {});

      expect(CallbackList.instance.count(), 1);
    });

    test('Callback List Removes Pending Entry', () {
      var calls = 0;

      CallbackList.instance.add("callback-removed", (_) => calls += 1);
      CallbackList.instance.remove("callback-removed");
      CallbackList.instance.execute("callback-removed", "");

      expect(calls, 0);
      expect(CallbackList.instance.count(), 0);
    });

    test('Callback List Ignores Unknown Key', () {
      CallbackList.instance.execute("callback-unknown", "");

      expect(CallbackList.instance.count(), 0);
    });

    test('Mapping List Clears', () {
      MappingList.instance.add(
        "sample.temporary",
        MappingItem((m, r) => r("")),
      );

      expect(MappingList.instance.has("sample.temporary"), true);

      MappingList.instance.clear();

      expect(MappingList.instance.has("sample.temporary"), false);
    });

    test('Unique ID Generates Distinct Keys', () {
      final keys = List.generate(1000, (_) => UniqueID.generate());

      expect(keys.toSet().length, 1000);
    });

    test('Mapping List Counts Entries', () {
      MappingList.instance.add("a", MappingItem((m, r) => r("")));
      MappingList.instance.add("b", MappingItem((m, r) => r("")));

      expect(MappingList.instance.count(), 2);
    });

    test('Message Returns Null For Another Type', () {
      final message = Message();
      message.set("number", 10);

      expect(message.get<int>("number"), 10);
      expect(message.get<String>("number"), null);
      expect(message.get<int>("missing"), null);
    });

    test('Empty Data View Reads As Nothing', () {
      // A view that carries no address describes nothing to read, and reading it must not take the isolate down.

      expect(ByteArrayHelper.createFromDataView(DataView(0, 16)).length, 0);
      expect(ByteArrayHelper.createFromDataView(DataView(128, 0)).length, 0);
      expect(ByteArrayHelper.createFromDataView(DataView(128, -1)).length, 0);
    });
    test('A Value That Is Not A Data View Is Answered With The Empty One', () {
      // The wire carries whatever the other side wrote, so a decoder answers rather than raising.

      final records = <String>[];
      final subscription = Logger(
        'XPLPC',
      ).onRecord.listen((record) => records.add(record.message));

      final view = DataView.fromJson(<String, dynamic>{'ptr': true, 'size': 1});

      subscription.cancel();

      expect(view.ptr, 0);
      expect(view.size, 0);
      expect(records.any((m) => m.contains('is not a data view')), true);
    });
  });
}
